File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        from typing import List

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        zero_count = nums.count(0)
        
        if zero_count > 1:
            return [0] * len(nums)
        
        elif zero_count == 1:
            total_product = 1
            for num in nums:
                if num != 0:
                    total_product *= num
            return [total_product if num == 0 else 0 for num in nums]
        
        else:
            total_product = 1
            for num in nums:
                total_product *= num
            return [total_product // num for num in nums]

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> v(nums.size(),1);
        int pre=1;
        for(int i=0;i<nums.size();i++){
            v[i]=pre;
            pre*=nums[i];
        }//v=[1,1,2,6]
        int suf=1;
        for(int i=nums.size()-1;i>=0;i--){
            v[i]=suf*v[i];
            suf*=nums[i];
        }//v=[24,12,8,6]
        return v;
    }
};